reference/async_reference/async_reference.cc

```diff
@@ -0,0 +1,238 @@
+
+		while(is_connected == true) {
+			auto async_err = tick_manager.validate_pending_options();
+
+			const auto sleep_duration = tick_rate - execution_duration;
+
+			auto wait_start = clock::now();
+			std::this_thread::sleep_for(sleep_duration - sleep_drift);
+			auto wait_end = clock::now();
+
+			sleep_drift =
+				sleep_duration - duration_cast<nanoseconds>(wait_start - wait_end);
+
+			if(async_err.error != ECSACT_ASYNC_OK) {
+				async_callbacks.add(async_err);
+
+				is_connected = false;
+				break;
+			}
+
+			auto start = clock::now();
+
+			auto cpp_options = tick_manager.move_and_increment_tick();
+
+			// TODO(Kelwan): Add done callbacks so we can resolve all requests
+			// https://github.com/ecsact-dev/ecsact_runtime/issues/102
+
+			ecsact_execution_events_collector collector;
+			collector.init_callback = &execution_callbacks::init_callback;
+			collector.update_callback = &execution_callbacks::update_callback;
+			collector.remove_callback = &execution_callbacks::remove_callback;
+
+			collector.init_callback_user_data = nullptr;
+			collector.update_callback_user_data = nullptr;
+			collector.remove_callback_user_data = nullptr;
+
+			std::unique_ptr<ecsact_execution_options> options = nullptr;
+
+			if(cpp_options) {
+				options = std::make_unique<ecsact_execution_options>(
+					util::cpp_to_c_execution_options(*cpp_options, *registry_id)
+				);
+			}
+
+			entity_manager.process_entities(async_callbacks, *registry_id);
+
+			auto systems_error =
+				ecsact_execute_systems(*registry_id, 1, options.get(), &collector);
+
+			auto end = clock::now();
+			execution_duration = duration_cast<nanoseconds>(end - start);
+
+			if(systems_error != ECSACT_EXEC_SYS_OK) {
+				async_callbacks.add(systems_error);
+				is_connected = false;
+				return;
+			}
+		}
+	});
+}
+
+void async_reference::flush_events(
+	const ecsact_execution_events_collector* execution_events,
+	const ecsact_async_events_collector*     async_events
+) {
+	async_callbacks.invoke(async_events);
+	if(registry_id) {
+		exec_callbacks.invoke(execution_events, *registry_id);
+	}
+}
+
+ecsact_async_request_id async_reference::create_entity_request() {
+	auto req_id = next_request_id();
+	if(is_connected == false && is_connected_notified == false) {
+		types::async_error async_err{
+			.error = ECSACT_ASYNC_ERR_PERMISSION_DENIED,
+			.request_ids = {req_id},
+		};
+
+		async_callbacks.add(async_err);
+		is_connected_notified = true;
+
+		return req_id;
+	}
+
+	entity_manager.request_entity(req_id);
+	return req_id;
+}
+
+void async_reference::disconnect() {
+	is_connected = false;
+	if(execution_thread.joinable()) {
+		execution_thread.join();
+	}
+}
+
+ecsact_async_request_id async_reference::next_request_id() {
+	return static_cast<ecsact_async_request_id>(
+		_last_request_id.fetch_add(1, std::memory_order_relaxed)
+	);
+}
+
+ecsact_async_request_id async_reference::convert_request_id(int32_t id) {
+	return static_cast<ecsact_async_request_id>(id);
+}
```

Comment:
this function should either be static or be a free function. it's not accessing any class state (`this`)